etna/reconciliation/top_down.py

```diff
@@ -0,0 +1,129 @@
+from enum import Enum
+
+import bottleneck as bn
+import pandas as pd
+from scipy.sparse import lil_matrix
+
+from etna.datasets import TSDataset
+from etna.reconciliation.base import BaseReconciliator
+
+
+class ReconciliationProportionsMethod(str, Enum):
+    """Enum for different reconciliation proportions methods."""
+
+    AHP = "AHP"
+    PHA = "PHA"
+
+    @classmethod
+    def _missing_(cls, method):
+        raise ValueError(
+            f"Unable to recognize reconciliation method '{method}'! "
+            f"Supported methods: {', '.join(sorted(m for m in cls))}."
+        )
+
+
+class TopDownReconciliator(BaseReconciliator):
+    """Top-down reconciliation methods."""
+
+    def __init__(self, target_level: str, source_level: str, period: int, method: str):
+        """Create top-down reconciliator from ``source_level`` to ``target_level``.
+
+        Parameters
+        ----------
+        target_level:
+            Level to be reconciled from the forecasts.
+        source_level:
+            Level to be forecasted.
+        period:
+            Period length for calculation reconciliation proportions.
+        method:
+            Proportions calculation method. Selects last ``period`` timestamps for estimation.
+            Currently supported options:
+
+            * AHP - Average historical proportions
+
+            * PHA - Proportions of the historical averages
+        """
+        super().__init__(target_level=target_level, source_level=source_level)
+
+        if period < 1:
+            raise ValueError("Period length must be positive!")
+
+        self.period = period
+        self.method = method
+
+        proportions_method = ReconciliationProportionsMethod(method)
+        if proportions_method == ReconciliationProportionsMethod.AHP:
+            self._proportions_method_func = self._estimate_ahp_proportion
+        elif proportions_method == ReconciliationProportionsMethod.PHA:
+            self._proportions_method_func = self._estimate_pha_proportion
+
+    def fit(self, ts: TSDataset) -> "TopDownReconciliator":
```

Comment:
Also check that values in `target` are non-negative, otherwise method might work incorrectly

Reply:
Added check and test